### PR TITLE
Fix enable OpenAPI Validator Plugin

### LINKED ISSUE #17: Refactor the openAPI validator plugin using new projectAPI
**Description:**
Refactor the openAPI plugin based on the new project API
- Start to implement the plugin using Syntax API 
- [x] Implement test for simple primitive record validation with json schema
- [x] Implement test for  record with array type attribute validate with json schema
- [x] Implement test for record with nested array type attribute validate with json schema
- [x] Implement test for  record with record type attribute validate with json schema
- [x] Implement test for record with nested record type attribute validate with json schema
- [x] Implement test for primitive validate with json schema

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### LINKED ISSUE #82: Enable OpenAPI Validator Plugin
**Description:**
Refactor openapi validator plugin

- [x] Extract annotation details for validation, To match consider the tags and operations provided with the annotation.

>Match available the openapi contract operation with the functions of the resource
Match available resources with the openapi contract operation

- [x] Implement a model for getting a summary of all the openapi contract path with paths' available methods and parameters.
- [x] Implement a model for getting a summary of all the resource function path with paths' available methods and parameters.
- [x] If the resource is missing service will be highlighted with the missing resource error 
- [ ] If there is an extra resource in service the resource will be highlighted with an error
>Implement the Service validator (with filtered operations)
- [x] Resource function node validates against the operation model
- [x] Operation model validate against the resource function node

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### LINKED ISSUE #83: Issue with multiple path parameters in openapi-to-ballerina command
**Description:**
When openapi contract documented with multiple path parameters, openapi-to-ballerina cli generated service including the last parameter as a path param type rest of the path parameters are not including path parameters.

**Steps to reproduce:**
`bal openapi -i <given openapi contract>`

- Used openapi contract
```opneapi
openapi: "3.0.0"
info:
  version: 1.0.0
  title: OpenApi Petstore
  license:
    name: MIT
servers:
  - url: http://petstore.{host}.io/v1
    description: The production API server
    variables:
      host:
        default: openapi
        description: this value is assigned by the service provider

tags:
  - name: pets
    description: Pets Tag
  - name: list
    description: List Tag

paths:
  /pets/{petId}/{petName}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      tags:
        - pets
      parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
        - name: petName
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string    
      responses:
        '200':
          description: Expected response to a valid request
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Pets"
        default:
          description: unexpected error
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Error"
components:
  schemas:
    Pet:
      required:
        - id
        - name
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        tag:
          type: string
        type:
          type: string
    Dog:
      allOf:
        - $ref: "#/components/schemas/Pet"
        - type: object
          properties:
            bark:
              type: boolean
    Pets:
      type: array
      items:
        $ref: "#/components/schemas/Pet"
    Error:
      required:
        - code
        - message
      properties:
        code:
          type: integer
          format: int32
        message:
          type: string
```
- Expected output
```ballerina
import ballerina/http;

listener http:Listener ep0 = new(80, config = {host: "petstore.openapi.io"});
service /v1 on ep0, ep1 {
# 
# + caller - Caller client object represents the endpoint
# + req    - Req represents the message, which came over the network
# + petId - The id of the pet to retrieve# + petName - The id of the pet to retrieve

# + return - Error value if an error occurred or return `()` otherwise
    resource function get pets/[string petId]/[string petName](http:Caller caller, http:Request req ) returns error? {
    }
}
```
- Actual output
```ballerina
import ballerina/http;


listener http:Listener ep0 = new(80, config = {host: "petstore.openapi.io"});

service /v1 on ep0, ep1 {
# 
# + caller - Caller client object represents the endpoint
# + req    - Req represents the message, which came over the network
# + petId - The id of the pet to retrieve# + petName - The id of the pet to retrieve

# + return - Error value if an error occurred or return `()` otherwise
    resource function get pets/{petId}/[string petName](http:Caller caller, http:Request req ) returns error? {

    }

}
```
 
**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**
swan-lake-alpha1

**OS, DB, other environment details and versions:**    

**Related Issues:**

### PR DESCRIPTION
**Pull request by lnash94**
Fix #17 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: openapi-cli/src/test/java/org/ballerinalang/openapi/CodeGeneratorTest.java [hunk 1]
```diff
@@ -147,6 +147,58 @@ public void generateSkeletonForRequestbody() {
...
+        final String serviceName = "openapipetstore";
+        String definitionPath = RES_DIR.resolve("multiPathParam.yaml").toString();;
+        CodeGenerator generator = new CodeGenerator();
+
+        try {
+            String expectedServiceContent = getStringFromGivenBalFile(expectedServiceFile, "generated_bal.bal");
+            generator.generateService(definitionPath, definitionPath, definitionPath, serviceName,
+                    resourcePath.toString(), filter);
+            if (Files.exists(resourcePath.resolve("openapipetstore-service.bal"))) {
+                String generatedService = getStringFromGivenBalFile(resourcePath, "openapipetstore-service.bal");
```
Comment (hemikak) [inline openapi-cli/src/test/java/org/ballerinalang/openapi/CodeGeneratorTest.java:161]: bal file names need to use underscore for word separation. `openapipetstore_service.bal`

File: openapi-validator/src/main/java/org/ballerinalang/openapi/validator/OpenAPIValidatorPlugin.java [hunk 2]
```diff
@@ -62,158 +46,20 @@ public void setCompilerContext(CompilerContext context) {
...
+    public List<Diagnostic> codeAnalyze(Project project) {
+        ServiceValidator serviceValidator = new ServiceValidator();
+        List<Diagnostic> diagnostics = new ArrayList<>();
+        try {
+            diagnostics = serviceValidator.validateResourceFunctions(project);
+            if (diagnostics.isEmpty()) {
+                return Collections.emptyList();
             }
+        } catch (IOException e) {
+            // handle proper
```
Comment (hemikak) [inline openapi-validator/src/main/java/org/ballerinalang/openapi/validator/OpenAPIValidatorPlugin.java:58]: Should we have to implement this ?

File: openapi-validator/src/main/java/org/ballerinalang/openapi/validator/OpenAPIValidatorPlugin.java [hunk 3]
```diff
@@ -62,158 +46,20 @@ public void setCompilerContext(CompilerContext context) {
...
+        } catch (OpenApiValidatorException e) {
+            // need to handle
```
Comment (hemikak) [inline openapi-validator/src/main/java/org/ballerinalang/openapi/validator/OpenAPIValidatorPlugin.java:60]: Should we have to implement this ?

File: openapi-validator/src/main/java/org/ballerinalang/openapi/validator/ResourceMethod.java [hunk 4]
```diff
@@ -31,15 +29,25 @@
...
         this.method = null;
         this.methodPosition = null;
         this.resourcePosition = null;
         this.parameters = new HashMap<>();
-        this.body = null;
+        this.body = false;
+    }
+
+    public Map<String, Node> getParameters() {
+
```
Comment (hemikak) [inline openapi-validator/src/main/java/org/ballerinalang/openapi/validator/ResourceMethod.java:44]: Can remove empty line.

File: openapi-validator/src/main/java/org/ballerinalang/openapi/validator/ResourcePathSummary.java [hunk 5]
```diff
@@ -34,7 +33,7 @@
     public ResourcePathSummary() {
         this.methods = new HashMap<>();
         this.path = null;
-        this.pathPosition = null;
+//        this.pathPosition = null;
```
Comment (hemikak) [inline openapi-validator/src/main/java/org/ballerinalang/openapi/validator/ResourcePathSummary.java:36]: Remove if not needed ?

File: openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/BaseTests.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                        Node node2 = typeNode.typeDescriptor();
+                    }
+                }
+                //-------------------------- comment due to service payload still
+                if (syntaxKind.equals(SyntaxKind.SERVICE_DECLARATION)) {
+                    ServiceDeclarationNode serviceDeclarationNode = (ServiceDeclarationNode) node;
+//                    ServiceBodyNode srvBNode = (ServiceBodyNode) serviceDeclarationNode.serviceBody();
+//                    // Extract Service Body for validate
+//                    int count = serviceDeclarationNode.internalNode().bucketCount();
+//                    STNode internalNode = srvBNode.internalNode();
```
Comment (hemikak) [inline openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/BaseTests.java:115]: Remove the commented code on this file ?

File: openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/Inputs.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package org.ballerinalang.openapi.validator.tests;
```
Comment (hemikak) [inline openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/Inputs.java:1]: Need copyright header

File: openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/Inputs.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+import io.ballerina.compiler.api.SemanticModel;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.syntax.tree.SyntaxTree;
+
+/**
+ * to-do.
+ */
```
Comment (hemikak) [inline openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/Inputs.java:9]: Update documentation

File: openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/Inputs.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+public class Inputs {
+    private   TypeSymbol paramType;
+    private   SemanticModel semanticModel;
+    private   SyntaxTree syntaxTree;
+
```
Comment (hemikak) [inline openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/Inputs.java:14]: File needs to be properly formatted

File: openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/RecordValidationTests.java [hunk 10]
```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (hemikak) [inline openapi-validator/src/test/java/org/ballerinalang/openapi/validator/tests/RecordValidationTests.java:2]: 2021

**Comment by lnash94** [comment]
Merged this PR to `wip_openapi_validator` because all the validation related implementations are working in progress due to some blockers.

**Pull request merged**